### PR TITLE
Fix traffic to terminating headless services

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/47348.

In previous versions, we used a different mechanism to get service instances that didn't filter unhealthy endpoints. We then switched to use the unified EDS source.

This fixes this by passing unhealthy endpoints through far enough to include them in SA computation, and filtering later.

This probably needs to be backported.

### CODE REVIEW

File: pilot/pkg/xds/eds_test.go [hunk 1]
```diff
@@ -352,171 +352,188 @@ func TestEDSOverlapping(t *testing.T) {
...
-		gotHealthy := sets.SortedList(sets.New(eh...))
-		gotUnhealthy := sets.SortedList(sets.New(euh...))
-		if !reflect.DeepEqual(gotHealthy, healthy) {
-			t.Fatalf("did not get expected endpoints: got %v, want %v", gotHealthy, healthy)
-		}
-		if !reflect.DeepEqual(gotUnhealthy, unhealthy) {
-			t.Fatalf("did not get expected unhealthy endpoints: got %v, want %v", gotUnhealthy, unhealthy)
-		}
-	}
+			t.Log("blah")
```
Comment (ramaraochavali) [inline pilot/pkg/xds/eds_test.go:412]: nit: remove this

File: pilot/pkg/model/endpointshards.go [hunk 2]
```diff
@@ -286,7 +286,7 @@ func (e *EndpointIndex) UpdateServiceEndpoints(
 	ep.Lock()
 	defer ep.Unlock()
 	newIstioEndpoints := istioEndpoints
-	if features.SendUnhealthyEndpoints.Load() {
+	if true {
```
Comment (ramaraochavali) [inline pilot/pkg/model/endpointshards.go:289]: remove this if?

File: releasenotes/notes/terminating-headless.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: traffic-management
+issue:
+  - 47348
+releaseNotes:
+  - |
+    **Fixed** an issue causing traffic to terminating headless service instances to not function correctly.
```
Comment (hzxuzhonghu) [inline releasenotes/notes/terminating-headless.yaml:8]: can you give more info, it occurs after a full push or just a eds

Comment (howardjohn) [reply releasenotes/notes/terminating-headless.yaml:8]: You cannot send to the terminating service, as its configuration is removed.

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**